---
title: "Database construction"
output: html_notebook
---

The general strategy for mapping the reads to functions is to first assemble contigs from the reads, then search those
contigs in the [M5nr database](https://doi.org/10.1186/1471-2105-13-141) and use those hits to map them by the M5 ids.

So, let's start by getting our mapping from contigs to M5nr ids (md5 hashes):

```{r}
library(mbtools)

contig_map <- read_blast("data/rnaseq/spades2m5nr.m8")[evalue < 1e-4]
```

## SEED database

We will now annotate our assembled transcripts via the mapping to the
M5NR database.

```{r}
seed <- fread("data/rnaseq/SEED.md52id2ont", header = FALSE)
seed[, V4 := NULL]
names(seed) <- c("reference", "id", "description")
```

Let's map the annotations to our contigs.

```{r}
seed <- seed[contig_map[, .(query, reference)], 
             on = "reference", nomatch = 0]
```

And we reduce to unique mappings.

```{r}
seed <- seed[, .(description = paste(unique(description), collapse=";")), 
             by = c("query", "id")]
seed[, uniqueN(query)] / contig_map[, uniqueN(query)]
```

Here we find almost 60% of the contigs. Finally, we combine that with the 
hierarchy to obtain the final mappings.

```{r}
hierarchy <- fread("data/rnaseq/SEED.id2subsystems", header = FALSE)
names(hierarchy) <- c("superpathway", "pathway", "subpathway", "function", "id")

seed <- hierarchy[seed, on="id"]
fwrite(seed, "data/rnaseq/transcripts_seed.csv")
```