---
title: "Differential transcript abundances for host data"
output: html_notebook
---

## Data reading and preprocesing

We will start by extracting the counts and connecting them to the sample data.

```{r}
library(mbtools)

txns <- fread("data/mouse_counts.csv")
txns[, sample := tstrsplit(sample, "-|_")[[3]]]
txns[, counts := round(counts)]
txns <- txns[counts > 0]
annotation <- unique(txns[, .(transcript, description)])
counts <- dcast(txns, transcript ~ sample, fill = 0, 
                fun.aggregate = sum, value.var = "counts")
nm <- counts[, transcript]
counts[, transcript := NULL]
counts <- as.matrix(counts)
rownames(counts) <- nm

meta <- fread("data/seq_samples.csv", key="Sample")[colnames(counts)]
meta[, day := substr(Sample, 4, 6)]
meta[, mouse := substr(Sample, 1, 3)]
meta
```

## Presence in samples

So how many mouse reads are there?

```{r}
txns[, sum(counts), by="sample"]
```

There are some reads in the sample but most transcripts are again specific for
individual samples. Very few transcripts are observed in the majorit of samples.

```{r, fig.width=6, fig.height=4}
ggplot(txns[, .N, by=transcript], aes(x=N)) + 
  geom_bar(stat = "count") + 
  labs(x = "number of samples", y = "transcript counts") + theme_bw()
```

To be consistent with the metatranscriptomics, we will also filter low abundance 
transcripts from the count matrix.

```{r}
counts <- t(filter_counts(t(counts), abundance = 10, presence = 0.25))
nrow(counts)
```

This leaves us with very few transcripts. Let's look at the samples in this
transcript space.

```{r, fig.width=6, fig.height=4}
pca <- prcomp(log2(mbtools::normalize(t(counts)) + 0.5))
explained <- pca$sdev^2 / sum(pca$sdev^2)
red <- as.data.table(pca$x)
red[, sample := rownames(pca$x)]
red <- meta[red[day == 29], on = c(Sample = "sample")]
ggplot(red, aes(x = PC1, y = PC2, shape = day, color = type)) + 
  geom_point(size = 2) +
  labs(x = sprintf("PC1 (%.1f%%)", explained[1] * 100), 
       y = sprintf("PC2 (%.1f%%)", explained[2] * 100)) +
  theme_bw()
ggsave("figures/mouse_txns_pca.svg", width = 6, height = 4)
```

There is no clear separation by transcript abundances. We can quantify this
with PERMANOVA again.

```{r}
library(vegan)

norm_counts <- log2(mbtools::normalize(t(counts[, meta$day >= 29])) + 0.5)
permanova <- adonis2(norm_counts ~ type, data = meta[meta$day >= 29], 
                     method = "euclidean")
permanova
```


## Differential tests

Let's see if there are individual mouse transcripts that distinguish the
phenotypes.

```{r, message = FALSE, warning = FALSE}
library(DESeq2)

setkey(meta, Sample)
meta <- meta[colnames(counts)]

days <- c(20, 25, 29) 
tests <- list()
for (d in days) {
  ds <- DESeqDataSetFromMatrix(counts[, meta$day == d], meta[day == d], ~ RIN + type)
  ds <- DESeq(ds, parallel = TRUE)
  res <- results(ds)
  res <- lfcShrink(ds, 3, parallel = TRUE)
  res$id <- rownames(res)
  coef_name <- resultsNames(ds)[3]
  res <- as.data.table(res)
  res[, day := d]
  res[, coefficient := coef_name]
  tests[[d]] <- res
}

tests <- rbindlist(tests)
```

We will merge this with the annotations for the contigs.

```{r}
tests <- annotation[tests, on = c(transcript = "id")]
fwrite(tests[order(padj)], "data/mouse_tests.csv")
```

However, looking for significant tests:

```{r}
tests[padj < 0.1]
```
there are none.

So nothing clearly differentiates the phenotype after antibiotic treatment
on the host transcript level.


