---
title: "Seaweed experiment 16s data"
author: "Christian Diener <cdiener@systemsbiology.org>"
output: html_notebook
---

# Effects of diet

## Data preparation

We will start by reading our data and required libraries.

```{r}
library(mbtools)

ps <- readRDS("data/seaweed_exp.rds")
```

We sill start by building up a dataframe that combines phyla abundances and sample metadata.

```{r}
phyla <- taxa_count(ps, "Phylum")

sdata <- as.data.table(as(sample_data(ps), "data.frame"))
phyla <- phyla[sdata, on = c(sample = "CategoricalID")]
head(phyla)
```

# Phyla abundances

## By exposure

First we have to assemble the relative abundances for controls.

```{r}
treated <- phyla[DaysElapsed == 31]
treated[, relative := reads / sum(reads, na.rm=T), 
        by = c("seaweed", "Response")]

high_taxa <- phyla[DaysElapsed == 31, 
                   .(taxa=taxa, rel=reads / sum(reads)), 
                   by = "sample"][rel > 1e-3, unique(taxa)] 
```

Let us visualize rough differences.

```{r, fig.height=4, fig.width=8}
pl <- ggplot(treated[taxa %in% high_taxa], 
             aes(x = factor(seaweed), y = relative, fill = taxa)) + 
  geom_bar(stat = "identity") + scale_y_continuous(labels = scales::percent) +
  facet_wrap(~ Response) + theme_bw() +
  scale_x_discrete(labels = c("without", "with")) +
  labs(x = "seaweed in diet", y = "relative abundance", fill = "phylum")
ggsave("figures/treatment_seaweed.png", height = 4, width = 8, dpi = 300)
pl
```

We can see that most of the Cyanobacteria and Proteobacteria are actually Chloroplasts and Mitochondria:

```{r}
non_responders <- subset_samples(ps, Response == "responder")
cyano_proteo <- subset_taxa(non_responders, 
                            Phylum %in% c("Cyanobacteria", "Proteobacteria"))
class <- taxa_count(cyano_proteo, "Order")[order(-reads), sum(reads), by = taxa]
fam <- taxa_count(cyano_proteo, "Family")[order(-reads), sum(reads), by = taxa]
print(class[, .(taxa, relative = V1 / sum(V1))])
print(fam[, .(taxa, relative = V1 / sum(V1))])
```

## Summarized

```{r, fig.width=4, fig.height=5}
library(gridExtra)

treated <- phyla[DaysElapsed == 31]
treated[, relative := reads / sum(reads), by = "Response"]

bars <- ggplot(treated[taxa %in% high_taxa], 
               aes(x = Response, y = relative, fill = taxa)) + 
  geom_bar(stat = "identity") + scale_y_continuous(labels = scales::percent) +
  theme_bw() + 
  theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1)) +
  labs(x = "", y = "relative abundance", fill = "phylum")
after_exposure <- sdata[DaysElapsed == 31]
resp_counts <- after_exposure[, table(Response, seaweed)][2:3, ]
tbl <- tableGrob(resp_counts)
pl <- grid.arrange(bars, tbl, nrow = 2, as.table = T, heights = c(3, 1))
ggsave("figures/summary_seaweed.svg", plot = pl, height = 6, width = 4)
pl
```

And the corresponding Fisher test:

```{r}
fisher.test(resp_counts) %>% print
```

# Beta diversity

We are interested in the beta diversity between samples after treatment, so we will start by selecting a subset of those
and calculating the ordination. We will look at the sequencing depth to get a decent depth for rarefaction.

```{r}
treat <- subset_samples(ps, DaysElapsed >= 20)
sample_sums(treat)
```

Okay so we can easily rarefy to 10k reads.

```{r, fig.width=6, fig.height=4}
ord <- rarefy_even_depth(treat, 5000) %>% 
       ordinate(., method = "PCoA", distance = "bray")

pl <- plot_ordination(treat, ord, color = "Response", 
                      shape = "TimepointCat") +
      stat_ellipse(aes(group = Response), type = "t") +
      scale_shape_manual(values = c(before = 1, during = 10, after = 19)) +
      labs(shape = "sample time") + theme_bw()
ggsave("figures/treatment_pcoa_seaweed.svg", width = 6, height = 4)
pl  
```

And the corresponding PERMANOVA

```{r}
library(vegan)

noresp <- subset_samples(treat, Response != "responder") %>% 
          rarefy_even_depth(., 5000)
counts <- as.matrix(taxa_count(noresp, NA))
sda <- as(sample_data(noresp), "data.frame")
permanova <- adonis(counts ~ seaweed*Response, data = sda)
permanova
```

So seaweed is not related to the response type.

To see which taxa mostly drive that:

```{r}
coef <- permanova$coefficients["Response1", ]
coeffs <- as(tax_table(ps)[names(coef), ], "matrix") %>% as.data.table
coeffs[, coef := coef]
coeffs[order(-abs(coef)), !"sequence", with = F]
```

# Relationship with qPCR cycle numbers

```{r, fig.height=4, fig.width=3}
pl <- ggplot(sdata[DaysElapsed == 31], aes(x = Response, y = 1 / Ct, 
                                           col = Response)) +
  geom_boxplot(width=0.5) + geom_jitter(width = 0.3) +
  guides(color = "none") + labs(x = "", y = "qPCR abundance [1/Ct]") +
  geom_hline(yintercept = 0.07, lty = "dashed") +
  theme_bw() + theme(axis.text.x = element_text(angle = 30, hjust = 1, 
                                                vjust = 1)) 
ggsave("figures/qpcr_cycles.svg", width = 3, height = 4)
pl
```

And we can see that there is a relationship between qPCR abundance estimates and response.

```{r}
pairwise.wilcox.test(sdata[, 1/Ct], sdata[, Response], 
                     p.adjust.method = "bonferroni")
```

We can check how that corresponds to the abundance of Mitochondria and Chloroplasts. For that we
will first get the relative abundances of those two in each sample.

```{r}
chloro_mit <- subset_taxa(ps, 
                          Class == "Chloroplast" | Family == "Mitochondria")
rel <- data.table(sample = sample_names(ps), 
                  relative = sample_sums(chloro_mit) / sample_sums(ps))
rel <- sdata[rel, on = c(CategoricalID = "sample")]
```

We can visualize this again.

```{r, fig.width=3, fig.height=4}
ggplot(rel, aes(x = TimepointCat, y = relative, color = Response)) +
  geom_hline(yintercept = 0.1, lty="dashed") +
  geom_jitter(width=0.2) +
  labs(x="sample time", y="Mitochondria/Chloroplasts [fraction of total]") +
  theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1))
ggsave("figures/seaweed_chloro_mit.svg", width=3, height=4)
```


Now we can check how that refers to cycles.

```{r, fig.height=4, fig.width=6}
pl <- ggplot(rel[DaysElapsed == 31], aes(x = 1/Ct, y = relative, col = Response)) + 
  scale_y_continuous(labels = scales::percent) + geom_point() +
  geom_vline(xintercept = 0.07, lty = "dashed") +
  geom_hline(yintercept = 0.05, lty = "dashed") +
  theme_bw() + labs(x = "qPCR abundance [1/Ct]", 
                    y = "Mitochondria or Chloroplasts")
ggsave("figures/qpcr_vs_chloro_mit.svg", height = 4, width = 6)
pl
```

Or rather do it with discretized cycle times.

```{r, fig.height=4, fig.width=4.5}
rel[, qpcr_material := c("low", "high")[(1/Ct > 0.07) + 1]]
pl <- ggplot(rel[DaysElapsed == 31], aes(x = Response, y = relative, 
                                         col = Response, shape = qpcr_material)) + 
  scale_y_continuous(labels = scales::percent) +
  geom_boxplot(position=position_dodge(width=0.6), width=0.5, outlier.color = NA) + 
  geom_jitter(width=0.3, aes(group = Response)) +
  geom_hline(yintercept = 0.05, lty = "dashed") +
  theme_bw() + theme(axis.text.x = element_text(angle = 30, hjust = 1, 
                                                vjust = 1)) +
  guides(color = "none") + labs(x = "", y = "Mitochondria or Chloroplasts" , 
                                shape = "qPCR abundance")
ggsave("figures/chloro_mit.svg", height = 4, width = 4.5)
pl
```

And there is also a relationship between the fraction of Chloroplasts and Mitochondria
and Response:

```{r}
pairwise.wilcox.test(rel[, relative], rel[, Response], 
                     p.adjust.method = "bonferroni")
```

## Weight

```{r}
weights <- fread("data/weights.csv")
names(weights)[1] <- "Subject"
weights <- melt(weights, id.vars = "Subject", 
                variable.name = "date", value.name = "weight")
weights[, date := as.Date(date, "%m/%d/%y")]
weights[, day := date - min(date) + 1]
weights <- weights[sdata[DaysElapsed == 31], on="Subject"]
weights[, weight_delta := weight - weight[day == 0], by="Subject"]
weights[, seaweed := c("no seaweed", "with seaweed")[seaweed + 1]]
weights
```

```{r, fig.width=7, fig.height=4}
pl <- ggplot(weights, aes(x=day, y=weight)) + 
  annotate("rect", xmin = 0, xmax = 20, ymin = 15, ymax = 23, 
           fill = "seagreen", alpha = 0.3) +
  annotate("rect", xmin = 25, xmax = 31, ymin = 15, ymax = 23, 
           fill = "tomato", alpha = 0.3) +
  geom_point(alpha = 0.5) + geom_line(aes(group = Subject), alpha = 0.5) + 
  facet_grid(seaweed ~ Response) + theme_bw() + ylim(15, 23) +
  labs(x = "time [days]", y = "weight [g]") + guides(color = "none")
ggsave("figures/weights_seaweed.svg", width = 7, height = 4)
pl
```

## Diet effect on diversity

## Shannon index

```{r, fig.width=6, fig.height=4}
controls <- subset_samples(ps, Response == "control") %>% 
            rarefy_even_depth(10000)
shannon <- estimate_richness(controls, measures = "Shannon")
shannon <- data.table(sample = rownames(shannon), shannon = shannon[, 1])
shannon <- sdata[shannon, on = c(CategoricalID = "sample")]

pl <- ggplot(shannon, 
             aes(x = factor(DaysElapsed), y = shannon, color = seaweed, 
                 group = interaction(DaysElapsed, seaweed))) + 
  geom_jitter(position = position_dodge(width = 0.5)) +
  theme_bw() + scale_color_discrete(labels = c("normal diet", "seaweed diet")) +
  labs(y = "alpha diversity [Shannon index]", x = "day", color = "")
ggsave("figures/shannon_seaweed_time.svg", width = 6, height = 4)
pl
```

Corresponding tests:

```{r}
shannon[, .(pval = wilcox.test(shannon[seaweed], shannon[!seaweed])$p.value), 
        by = "DaysElapsed"]
```

Phyla distributions:

```{r, fig.height = 5, fig.width = 8}
phyla[, relative := reads / sum(reads), by = Sample]
pl <- ggplot(phyla[taxa %in% high_taxa], 
             aes(x = factor(DaysElapsed), y = reads, fill = taxa)) + 
  geom_bar(stat = "identity", position = "fill") + 
  scale_y_continuous(labels = scales::percent) +
  facet_grid(c("normal diet", "seaweed diet")[as.numeric(seaweed) + 1] ~ Response) + 
  theme_classic() +
  labs(x = "day", y = "relative abundance", fill = "phylum")
ggsave("figures/phyla_day_seaweed.png", height = 5, width = 8, dpi = 300)
pl
```

And the corresponding PERMANOVA:

```{r}
counts <- as.matrix(taxa_count(controls, NA))
sda <- as(sample_data(controls), "data.frame")
permanova <- adonis(counts ~ seaweed, data = sda)
permanova
```