---
title: "Differential transcript abundances"
output: html_notebook
---

## Data reading and preprocesing

```{r}
library(mbtools)

counts <- fread("data/rnaseq/txcounts.csv")
counts[, sample := tstrsplit(sample, "-")[[3]]]
counts[, counts := round(counts)]
counts[counts > 0]
counts <- dcast(counts, transcript ~ sample, fill=0, value.var = "counts")
nm <- counts[, transcript]
counts[, transcript := NULL]
counts <- as.matrix(counts)
rownames(counts) <- nm

meta <- fread("data/rnaseq/seq_samples.csv", key="Sample")[colnames(counts)]
meta[, day := substr(Sample, 4, 6)]
meta[, mouse := substr(Sample, 1, 3)]
meta
```

## Contig metrics

We will start with some general metrics for the assembled transcripts.

```{r}
library(stringr)

node_ex <- "NODE_(\\d+)_length_(\\d+)_cov_([\\d\\.]+)_g(\\d+)_i(\\d+)"
contigs <- as.data.table(str_match(rownames(counts), node_ex))
names(contigs) <- c("id", "node", "length", "kmer_coverage", "gene", "isoform")
contigs[, length := as.numeric(length)]
contigs[, kmer_coverage := as.numeric(kmer_coverage)]
contigs[, count := rowSums(counts > 0)]
```

transcript lengths:

```{r, fig.width=6, fig.height=4}
ggplot(contigs, aes(x = length)) + 
  geom_histogram(bins = 50) + 
  scale_x_log10() + 
  labs(x="transcript length [bps]") + theme_bw()
ggsave("figures/transcript_length.svg", width = 6, height = 4)
```

transcript coverage:

```{r, fig.width=6, fig.height=4}
ggplot(contigs, aes(x = kmer_coverage)) + 
  geom_histogram(bins = 50) + scale_x_log10() + 
  labs(x = "transcript kmer coverage") + theme_bw()
ggsave("figures/transcript_cov.svg", width = 6, height = 4)
```

presence in samples:

```{r, fig.width=6, fig.height=4}
ggplot(contigs, aes(x=factor(count))) + 
  geom_bar(stat = "count") + 
  labs(x = "number of samples", y = "contig counts") + theme_bw()
ggsave("figures/contig_counts.svg", width = 6, height = 4)
```

We will also filter low abundance transcripts from the count matrix.

```{r}
counts <- t(filter_counts(t(counts), abundance = 10, presence = 0.25))
nrow(counts)
```

Let's look at the samples.

```{r, fig.width=6, fig.height=4}
pca <- prcomp(log2(mbtools::normalize(t(counts)) + 0.5))
explained <- pca$sdev^2 / sum(pca$sdev^2)
red <- as.data.table(pca$x)
red[, sample := rownames(pca$x)]
red <- meta[red, on = c(Sample = "sample")]
ggplot(red[day == 29], aes(x = PC1, y = PC2, shape = day, color = type)) + 
  geom_point(size = 2) +
  labs(x = sprintf("PC1 (%.1f%%)", explained[1] * 100), 
       y = sprintf("PC2 (%.1f%%)", explained[2] * 100)) +
  theme_bw()
ggsave("figures/transcript_pca.svg", width = 6, height = 4)
```

## With responders

```{r, message = FALSE, warning = FALSE}
library(DESeq2)

setkey(meta, Sample)
meta <- meta[colnames(counts)]

days <- c(20, 25, 29) 
tests <- list()
for (d in days) {
  ds <- DESeqDataSetFromMatrix(counts[, meta$day == d], meta[day == d], ~ RIN + type)
  ds <- DESeq(ds, parallel = TRUE)
  res <- results(ds)
  res <- lfcShrink(ds, 3, parallel = TRUE)
  res$id <- rownames(res)
  coef_name <- resultsNames(ds)[3]
  res <- as.data.table(res)
  res[, day := d]
  res[, coefficient := coef_name]
  tests[[d]] <- res
}

tests <- rbindlist(tests)
```

We will merge this with the annotations for the contigs.

```{r}
anns <- fread("data/rnaseq/transcripts_seed.csv")
tests <- anns[tests, on = c(query = "id")]
sig <- tests[padj < 0.05][order(padj)]
```

However if we compare the dimensions:

```{r}
tests[day == 29, table(padj < 0.05)]
```

So more than a third of all tests are significant. We can also check that by day.

```{r}
tests[, sum(padj < 0.05, na.rm = T) / sum(!is.na(padj)), by = c("day", "coef_name")]
```

This is still below the nominal p-value, but the transcripts themselves are sample specific. 